### PR TITLE
Add Windows unencrypted credentials option

### PR DESCRIPTION
**Pull request by confusdcodr**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -19,17 +19,16 @@ supporting pillar.
...
-      key:
-      oupath:
-      admin_users:
-      admin_groups:
-      ec2config:
-      tries:
+    dns_name:
+    netbios_name:
+    username:
+    encrypted_password:
```
Comment (userhas404d) [inline README.md]: should probably add a note here about `rotated_password`

File: join-domain/windows/files/JoinDomain.ps1 [hunk 2]
```diff
@@ -13,19 +13,24 @@ Param(
...
     $Key,
 
-    [parameter(Mandatory = $true)]
+    [parameter(Mandatory = $false)]
     [String]
         #Encrypted join domain password
     $EncryptedPassword,
 
-    [parameter(Mandatory = $true)]
+    [parameter(Mandatory = $false)]
```
Comment (lorengordon) [inline join-domain/windows/files/JoinDomain.ps1]: Use powershell Parameter Sets to make `Credential` and `Key`/`EncryptedPassword` mutually exclusive. Then remove the `if`/`throw` test.

https://devblogs.microsoft.com/scripting/use-parameter-sets-to-simplify-powershell-commands/

File: join-domain/windows/init.sls [hunk 3]
```diff
@@ -6,8 +6,9 @@ join standalone system to domain:
     - args: >-
         -DomainName "{{ join_domain.dns_name }}"
         -TargetOU "{{ join_domain.oupath }}"
-        -Key "{{ join_domain.key }}"
-        -EncryptedPassword "{{ join_domain.encrypted_password }}"
+        -RotatedPassword "{{ join_domain.get("rotated_password", "") }}"
```
Comment (lorengordon) [inline join-domain/windows/init.sls]: Use a jinja `if` to pass _only_ `Credential` or `Key`/`EncryptedPassword`

File: README.md [hunk 4]
```diff
@@ -19,17 +19,16 @@ supporting pillar.
 ```yaml
 join-domain:
   lookup:
-    windows:
```
Comment (lorengordon) [inline README.md]: The header here is `join-domain:windows` (indicating the sub-key being removed), which needs to be updated as well. This was a relic of an older implementation...

File: join-domain/windows/init.sls (hunk 3 above)
Comment (lorengordon) [reply join-domain/windows/init.sls]: Also, change the pillar key to `password` because the rotation aspect is an implementation detail that is not specifically related to this formula

File: README.md (hunk 4 above)
Comment (confusdcodr) [reply README.md]: So is this indication of the subkey needing to be removed just a salt thing I am not aware of? I only figured out I needed to remove this subkey when testing by looking at the pillar.example

File: join-domain/windows/init.sls (hunk 3 above)
Comment (confusdcodr) [reply join-domain/windows/init.sls]: So I intentionally used the get function instead of `if/else` because the spacing that occurs when the command is ultimately executed confuses powershell and it will no longer recognize that the parameter is there. I have also attempted to trim the spacing using the jinja dash `{%-` but that interferes with the previous parameter

File: join-domain/windows/files/JoinDomain.ps1 [hunk 5]
```diff
@@ -567,13 +572,13 @@ Function xAdd-Computer
     )
 
     #wrapper for the Add-Computer cmdlet that adds an OU conditional
-    if($TargetOU  -eq "" -and $TargetOU -eq [String]::Empty)
+    if($TargetOU -eq "" -and $TargetOU -eq [String]::Empty -or $TargetOU -eq "None")
```
Comment (lorengordon) [inline join-domain/windows/files/JoinDomain.ps1]: This change seems unrelated. What is the purpose here?

File: join-domain/windows/files/JoinDomain.ps1 [hunk 6]
```diff
@@ -584,13 +589,22 @@ $DomainJoinStatus = Get-DomainJoinStatus -DomainFQDN $DomainName
...
 {
     #If the object is found, remove it, else add the computer to the domain.
-    #Create the credential
-    $AesObject = New-Object System.Security.Cryptography.AesCryptoServiceProvider
-    $AesObject.IV = New-Object Byte[]($AesObject.IV.Length)
-    $AesObject.Key = [System.Convert]::FromBase64String($Key)
-    $EncryptedStringBytes = [System.Convert]::FromBase64String($EncryptedPassword)
-    $ReEncryptedPassword = ConvertTo-SecureString -String "$([System.Text.UnicodeEncoding]::Unicode.GetString(($AesObject.CreateDecryptor()).TransformFinalBlock($EncryptedStringBytes, 0, $EncryptedStringBytes.Length)))" -AsPlainText -Force
-    $cred = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList $UserName, $ReEncryptedPassword
+    #Create the credential from the auto-rotate password when it's provided to the formula
```
Comment (lorengordon) [inline join-domain/windows/files/JoinDomain.ps1]:
```suggestion
    #Create the credential from the Password parameter set
```

File: join-domain/windows/files/JoinDomain.ps1 [hunk 7]
```diff
@@ -584,13 +589,22 @@ $DomainJoinStatus = Get-DomainJoinStatus -DomainFQDN $DomainName
...
+    if($Password)
```
Comment (lorengordon) [inline join-domain/windows/files/JoinDomain.ps1]: This would probably be better handled using a [`switch` statement](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_switch?view=powershell-7)
```suggestion
    switch ($PSCmdlet.ParameterSetName)
    {
        "Password" { ... }
        "EncryptedPassword" { ... }
    }
```

File: join-domain/windows/files/JoinDomain.ps1 [hunk 8]
```diff
@@ -13,19 +13,24 @@ Param(
...
+        #Username of account used to domain join the target computer
+    $UserName,
+
+    [parameter(ParameterSetName='EncryptedPassword')]
     [String]
         #Key used to decrypt the join domain key
     $Key,
 
-    [parameter(Mandatory = $true)]
+    [parameter(ParameterSetName='EncryptedPassword')]
```
Comment (lorengordon) [inline join-domain/windows/files/JoinDomain.ps1]:
```suggestion
    [parameter(Mandatory = $true, ParameterSetName = "EncryptedPassword")]
```

File: join-domain/windows/files/JoinDomain.ps1 [hunk 9]
```diff
@@ -13,19 +13,24 @@ Param(
...
     [String]
         #Encrypted join domain password
     $EncryptedPassword,
 
-    [parameter(Mandatory = $true)]
+    [parameter(ParameterSetName='Password')]
```
Comment (lorengordon) [inline join-domain/windows/files/JoinDomain.ps1]:
```suggestion
    [parameter(Mandatory = $true, ParameterSetName = "Password")]
```

File: join-domain/windows/files/JoinDomain.ps1 [hunk 10]
```diff
@@ -13,19 +13,24 @@ Param(
     $TargetOU = "null",
 
     [parameter(Mandatory = $true)]
+    [String]
+        #Username of account used to domain join the target computer
+    $UserName,
+
+    [parameter(ParameterSetName='EncryptedPassword')]
```
Comment (lorengordon) [inline join-domain/windows/files/JoinDomain.ps1]:
```suggestion
    [parameter(Mandatory = $true, ParameterSetName = "EncryptedPassword")]
```

File: join-domain/windows/init.sls [hunk 11]
```diff
@@ -1,18 +1,31 @@
...
+    - source: {{ join_domain.script.source }}
+    - makedirs: true
+
 join standalone system to domain:
-  cmd.script:
-    - name: salt://{{ tpldir }}/files/JoinDomain.ps1
-    - args: >-
+  cmd.run:
+    - name: >-
+        powershell "{{ join_domain.script.name }}"
```
Comment (lorengordon) [inline join-domain/windows/init.sls]: The `shell: powershell` line handles the powershell invocation. This just needs to call the script.

```suggestion
        & "{{ join_domain.script.name }}"
```

File: README.md [hunk 12]
```diff
@@ -14,22 +14,22 @@ This formula uses data externalized via the SaltStack "[Pillar](https://docs.sal
...
-      key:
-      oupath:
-      admin_users:
-      admin_groups:
-      ec2config:
-      tries:
+    dns_name:
+    netbios_name:
+    username:
+    encrypted_password:
```
Comment (lorengordon) [inline README.md]: Let's apply the same treatment here as in pillar.example, with the mutually exclusive commentary...

File: join-domain/windows/files/JoinDomain.ps1 (hunk 5 above)
Comment (confusdcodr) [reply join-domain/windows/files/JoinDomain.ps1]: When you don't define TargetOU in the pillar file, it comes through as "None" instead of as "" which is what the original if statement believed would happen.

File: join-domain/windows/files/JoinDomain.ps1 (hunk 7 above)
Comment (confusdcodr) [reply join-domain/windows/files/JoinDomain.ps1]: Why use a switch statement if I am only checking a single condition? This seems like extra nesting for no reason.

File: join-domain/windows/init.sls (hunk 11 above)
Comment (confusdcodr) [reply join-domain/windows/init.sls]: I explicitly put in powershell instead of & as the `&` option will run it in the background and suppress any errors from surfacing.

Comment (lorengordon) [reply join-domain/windows/init.sls]: This needs some proof. The `-ErrorAction Stop` argument should force an errexit, for the ps script. Salt already calls this with `Powershell -NonInteractive -NoProfile`. Invoking another powershell process should be wholly unnecessary.

Comment (lorengordon) [reply join-domain/windows/init.sls]: here is an example of a success and a forced failure, using `cmd.run` and `& <script>`:

```
C:\WINDOWS\system32> C:\salt\salt-call.bat state.sls foo
local:
----------
          ID: run foo.ps1
    Function: cmd.run
        Name: & "c:/salt/srv/salt/foo/foo.ps1" -SecureString (ConvertTo-SecureString -String "i like cheese" -AsPlainText -Force) -ErrorAction Stop
      Result: True
     Comment: Command "& "c:/salt/srv/salt/foo/foo.ps1" -SecureString (ConvertTo-SecureString -String "i like cheese" -AsPlainText -Force) -ErrorAction Stop" run
     Started: 18:30:47.222000
    Duration: 425.0 ms
     Changes:
              ----------
              pid:
                  15064
              retcode:
                  0
              stderr:
              stdout:
                  i like cheese

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time: 425.000 ms
C:\WINDOWS\system32> C:\salt\salt-call.bat state.sls foo
[ERROR   ] Command 'chcp 437 > nul & Powershell -NonInteractive -NoProfile "& \"c:/salt/srv/salt/foo/foo.ps1\" -SecureString (ConvertTo-SecureString -String \"i like cheese\" -AsPlainText -Force) -ErrorAction Stop"' failed with return code: 1
[ERROR   ] stderr: ScriptHalted
At C:\salt\srv\salt\foo\foo.ps1:6 char:1
+ throw
+ ~~~~~
    + CategoryInfo          : OperationStopped: (:) [], RuntimeException
    + FullyQualifiedErrorId : ScriptHalted
[ERROR   ] retcode: 1
[ERROR   ] {u'pid': 24320, u'retcode': 1, u'stderr': u'ScriptHalted\r\nAt C:\\salt\\srv\\salt\\foo\\foo.ps1:6 char:1\r\n+ throw\r\n+ ~~~~~\r\n    + CategoryInfo          : OperationStopped: (:) [], RuntimeException\r\n    + FullyQualifiedErrorId : ScriptHalted', u'stdout': u''}
local:
----------
          ID: run foo.ps1
    Function: cmd.run
        Name: & "c:/salt/srv/salt/foo/foo.ps1" -SecureString (ConvertTo-SecureString -String "i like cheese" -AsPlainText -Force) -ErrorAction Stop
      Result: False
     Comment: Command "& "c:/salt/srv/salt/foo/foo.ps1" -SecureString (ConvertTo-SecureString -String "i like cheese" -AsPlainText -Force) -ErrorAction Stop" run
     Started: 18:31:16.711000
    Duration: 517.0 ms
     Changes:
              ----------
              pid:
                  24320
              retcode:
                  1
              stderr:
                  ScriptHalted
                  At C:\salt\srv\salt\foo\foo.ps1:6 char:1
                  + throw
                  + ~~~~~
                      + CategoryInfo          : OperationStopped: (:) [], RuntimeException
                      + FullyQualifiedErrorId : ScriptHalted
              stdout:

Summary for local
------------
Succeeded: 0 (changed=1)
Failed:    1
------------
Total states run:     1
Total run time: 517.000 ms
```

File: join-domain/windows/files/JoinDomain.ps1 (hunk 7 above)
Comment (lorengordon) [reply join-domain/windows/files/JoinDomain.ps1]: It's not a single condition. It's actually two. One being `Password`, the other being `EncryptedPassword`. Using a `switch` is a setup for supporting multiple cases in an approach that is more declarative than imperative. It's not just about nesting.

But either way, the condition to test should be the ParamterSetName

```suggestion
    if($PSCmdlet.ParameterSetName)
```

File: join-domain/windows/files/JoinDomain.ps1 (hunk 5 above)
Comment (lorengordon) [reply join-domain/windows/files/JoinDomain.ps1]: I'm not sure what would be bringing it in as "None". That definitely seems like a separate issue that should be investigated more in its own right.

Comment (confusdcodr) [reply join-domain/windows/files/JoinDomain.ps1]: salt/python is bringing it in as none. If I don't put the field in the pillar at all then it will come in as "" but if you define the field but provide no value it comes in as "None". Anywho, I can remove this addition as there is no reason for the field to be defined if you aren't providing a value.

File: join-domain/windows/files/JoinDomain.ps1 [hunk 13]
```diff
@@ -584,13 +589,22 @@ $DomainJoinStatus = Get-DomainJoinStatus -DomainFQDN $DomainName
...
-    $cred = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList $UserName, $ReEncryptedPassword
+    #Create the credential from the Password parameter set
+    if($PSCmdlet.ParameterSetName -eq "Password")
+    {
+        $SecPassword = ConvertTo-SecureString -String $Password -AsPlainText -Force
+        $cred = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList $UserName, $SecPassword
+    }
+    else
+    {
+        #Create the credential from key and encrypted password when they're provided to the formula
```
Comment (lorengordon) [inline join-domain/windows/files/JoinDomain.ps1]:
```suggestion
        #Create the credential from the EncryptedPassword parameter set
```

File: join-domain/elx/map.jinja [hunk 14]
```diff
@@ -13,7 +13,7 @@
 
 {%- set admin_groups = [] %}
 {%- for group in join_domain.get('admin_groups', []) +
-                 join_domain_grains.get('admin_groups', [])
+                join_domain_grains.get('admin_groups', [])
```
Comment (lorengordon) [inline join-domain/elx/map.jinja]: What's causing this change?

Comment (confusdcodr) [reply join-domain/elx/map.jinja]: EditorConfig... I ran it against a few files.
